cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java

```diff
@@ -336,6 +340,93 @@ public void apply() throws Exception {
     }
   }
 
+  /** Method to add version in JobKey and VersionKey in SchedulerStore.
+   *
+   * @throws Exception
+   */
+  public void upgrade() throws Exception {
+    // Create table instance.
+    initializeScheduleTable();
+    factory.createExecutor(ImmutableList.of((TransactionAware) table))
+      .execute(new TransactionExecutor.Subroutine() {
+        @Override
+        public void apply() throws Exception {
+          upgradeJobs();
+          upgradeTriggers();
+        }
+      });
+  }
+
+  private void upgradeJobs() throws Exception {
+    Row result = table.get(JOB_KEY);
+
+    if (result.isEmpty()) {
+      return;
+    }
+
+    for (byte[] column : result.getColumns().values()) {
+      JobDetail jobDetail = (JobDetail) SerializationUtils.deserialize(column);
+      JobKey oldJobKey = jobDetail.getKey();
+      String oldJobKeyName = oldJobKey.getName();
+      String[] splits = oldJobKeyName.split(":");
+      // Old JobKey name has format = namespace:application:type:program
+      if (splits.length != 4) {
+        LOG.warn("Skip upgrading Job {} , because its key does not have 5 parts", oldJobKey);
+        continue;
+      }
+
+      // New JobKey name has format = namespace:application:version:type:program
+      String newJobName = getNameWithVersion(splits);
+
+      JobDetail newJobDetail = jobDetail.getJobBuilder().withIdentity(newJobName, oldJobKey.getGroup()).build();
+      persistJob(table, newJobDetail);
+      removeJob(table, oldJobKey);
+    }
+  }
+
+  private void upgradeTriggers() throws Exception {
+    Row result = table.get(TRIGGER_KEY);
+    if (result.isEmpty()) {
+      return;
+    }
+
+    for (byte[] column : result.getColumns().values()) {
+      TriggerStatusV2 triggerStatus = (TriggerStatusV2) SerializationUtils.deserialize(column);
+      TriggerKey oldTriggerKey = triggerStatus.trigger.getKey();
+      JobKey oldTriggerJobKey = triggerStatus.trigger.getJobKey();
+
+      String[] splits = oldTriggerKey.getName().split(":");
+      // Old TriggerKey name has format = namespace:application:type:program:schedule
+      if (splits.length != 5) {
+        LOG.warn("Skip upgrading Trigger {} , because its key does not have 5 parts", oldTriggerKey);
```

Comment:
What do mean by `does not have 5 parts`? This warning message is not very useful. You should have something more meaningful, like `("Unexpected key format {}. Expected in format 'namespace:application:type:program:schedule'", oldTriggerKey)`